File: csharp/src/Solutions.Lib/0001_TwoSum/TwoSumSolutionC.cs
namespace Solutions.Lib;

/// <summary>
/// A cleaner Two-pass Hash Table solution
/// </summary>
public class TwoSumSolutionC : TwoSumSolution
{
	// Time Complexity: O( N )
	// Space Complexity: O( N )
	protected override int[] TwoSum(int[] nums, int target)
	{
		int count = nums.Length;
		Dictionary<int, int> map = new();

		// map all of the nums to their locations
		for (int i = 0; i < count; i++)
		{
			int num = nums[i];
			map[num] = i;
		}

		// search for a num complement in the array
		for (int i = 0; i < count; i++)
		{
			int num = nums[i];
			int complement = target - num;

			// making use of our quick lookup map
			if (map.ContainsKey(complement))
			{
				int complementIndex = map[complement];

				if (i != complementIndex)
				{
					return new int[] { complementIndex, i };
				}
			}
		}

		return new int[] { count - 2, count - 1 };
	}
}

File: javascript/src/0001_TwoSum/TwoSumSolutionC.js
import { BaseTwoSumSolution } from './BaseTwoSumSolution.js';

// Brute Force Solution
export class TwoSumSolutionC extends BaseTwoSumSolution {
	// Time Complexity: O( N )
	// Space Complexity: O( N )
	twoSum(nums, target) {
		const count = nums.length;
		const map = new Map();

		// map all of the nums to their locations
		for (let i = 0; i < count; i++)
		{
			const num = nums[i];
			map.set(num, i);
		}

		// search for a num complement in the array
		for (let i = 0; i < count; i++)
		{
			const num = nums[i];
			const complement = target - num;

			// making use of our quick lookup map
			if (map.has(complement)) {
				const complementIndex = map.get(complement);

				if (i !== complementIndex) {
					return [i, complementIndex];
				}
			}
		}

		// this line should never be reached
		return [];
	}
}
